src/Components/Server/src/Circuits/ServerComponentDeserializer.cs

```diff
@@ -0,0 +1,254 @@
+            {
+                if (marker.Type != ServerComponentMarker.ServerMarkerType)
+                {
+                    Log.InvalidMarkerType(_logger, marker.Type);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Descriptor == null)
+                {
+                    Log.MissingMarkerDescriptor(_logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                var (descriptor, serverComponent) = DeserializeServerComponent(marker);
+                if (descriptor == null)
+                {
+                    // We failed to deserialize the component descriptor for some reason.
+                    descriptors.Clear();
+                    return false;
+                }
+
+                // We force our client to send the descriptors in order so that we do minimal work.
+                // The list of descriptors starts with 0 and lastSequence is initialized to -1 so this
+                // check covers that the sequence starts by 0.
+                if (lastSequence != serverComponent.Sequence - 1)
+                {
+                    if (lastSequence == -1)
+                    {
+                        Log.DescriptorSequenceMustStartAtZero(_logger, serverComponent.Sequence);
+                    }
+                    else
+                    {
+                        Log.OutOfSequenceDescriptor(_logger, lastSequence, serverComponent.Sequence);
+                    }
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (lastSequence != -1 && !previousInstance.InvocationId.Equals(serverComponent.InvocationId))
+                {
+                    Log.MismatchedInvocationId(_logger, previousInstance.InvocationId.ToString("N"), serverComponent.InvocationId.ToString("N"));
+                    return false;
+                }
+
+                // As described below, we build a chain of descriptors to prevent being flooded by
+                // descriptors from a client not behaving properly.
+                lastSequence = serverComponent.Sequence;
+                previousInstance = serverComponent;
+                descriptors.Add(descriptor);
+            }
+
+            return true;
+        }
+
+        private (ComponentDescriptor, ServerComponent) DeserializeServerComponent(ServerComponentMarker record)
+        {
+            string unprotected;
+            try
+            {
+                unprotected = _dataProtector.Unprotect(record.Descriptor);
+            }
+            catch (Exception e)
+            {
+                Log.FailedToUnprotectDescriptor(_logger, e);
+                return default;
+            }
+
+            ServerComponent descriptorInstance;
+            try
+            {
+                descriptorInstance = JsonSerializer.Deserialize<ServerComponent>(
```

Comment:
Can we pick a better name than this? there's descriptor and descriptor instance again.

Reply:
Yes, it should be serverComponent. There is a bit of leftover from the rename

Reply:
updated the names